code/dc2/pad_control.cc

```diff
@@ -0,0 +1,87 @@
+#include "common/debug.h"
+#include "common/log.h"
+
+#include "dc2/pad_control.h"
+
+#include "host/host_interface_dwm.h"
+
+set_log_channel("pad_control");
+
+// 002ED3E0
+// Clears all recorded actions
+void CPadControl::Reset()
+{
+  log_trace("CPadControl::{}()", __func__);
+
+  m_buttons.clear();
+  m_analog.clear();
+}
+
+// 002ED460
+// Registers up to two buttons for a particular action
+// TODO: Accept a vector for an arbitrary number of buttons
+void CPadControl::RegisterBtn(const std::string_view& action_name, host::pad_handler::buttons btn1, host::pad_handler::buttons btn2)
+{
+  log_trace("CPadControl::{}({}, {:#X}, {:#X})", __func__, action_name, std::to_underlying(btn1), std::to_underlying(btn2));
+
+  m_buttons[action_name.data()] = {
+    .m_value = false,
+    .m_input_key = btn1 | btn2
+  };
+}
+
+// 002ED4A0
+// Registers an axis for a particular aciton
+void CPadControl::RegisterAnalog(const std::string_view& action_name, EPadAxis axis)
+{
+  log_trace("CPadControl::{}({}, {})", __func__, action_name, std::to_underlying(axis));
+
+  m_analog[action_name.data()] = {
+    .m_value = 0.0f,
+    .m_input_axis = axis
+  };
+}
+
+// 002ED4E0
+// Checks the button value for a particular action
+bool CPadControl::Btn(const std::string_view& action_name)
+{
+  log_trace("CPadControl::{}({})", __func__, action_name);
+
+  if (!m_buttons.contains(action_name.data()))
+  {
+    return false;
+  }
+
+  return m_buttons[action_name.data()].m_value;
```

Comment:
Probably not the best way to do this. `contains` and `operator[]` both have best case constant time but worst case linear. You can do both these operations at once with `find` and iterators.